scripts/language_model/word_language_model.py

```diff
@@ -354,30 +357,51 @@ def train():
             data_list = gluon.utils.split_and_load(data, context, batch_axis=1, even_split=True)
             target_list = gluon.utils.split_and_load(target, context, batch_axis=1, even_split=True)
             hiddens = detach(hiddens)
+            model.collect_params().zero_grad()
             Ls = []
             L = 0
             with autograd.record():
                 for j, (X, y, h) in enumerate(zip(data_list, target_list, hiddens)):
-                    output, h, encoder_hs, dropped_encoder_hs = model.forward(X, h)
+                    output, h, encoder_hs, dropped_encoder_hs = model(X, h)
                     l = joint_loss(output, y, encoder_hs, dropped_encoder_hs)
                     L = L + l.as_in_context(context[0]) / X.size
-                    Ls.append(l/X.size)
+                    Ls.append(l / X.size)
                     hiddens[j] = h
             L.backward()
             grads = [p.grad(d.context) for p in parameters for d in data_list]
             gluon.utils.clip_global_norm(grads, args.clip)
 
+            param_dict_batch_i = {k: v.data(context[0]).copy()
+                                  for k, v in model.collect_params().items()}
             trainer.step(1)
 
+            alpha = 1.0 / max(1, batch_i - avg_trigger + 1)
+            if param_dict_avg is None:
+                param_dict_avg = {k: v.data(context[0]).copy()
+                                  for k, v in model.collect_params().items()}
+            for name, param_avg in param_dict_avg.items():
+                param_avg[:] += alpha * (param_dict_batch_i[name] - param_avg)
```

Comment:
We should move this averaging step after the validation, as avg_trigger is determined after that